File: demo-app-objc/CloudXObjCRemotePods/Views/Ad/InterstitialViewController.m
#import "InterstitialViewController.h"
#import <CloudXCore/CloudXCore.h>
#import "DemoAppLogger.h"
#import "CLXDemoConfigManager.h"
#import "UserDefaultsSettings.h"
#import "NSError+DemoDescription.h"

@interface InterstitialViewController ()
@property (nonatomic, strong) CLXInterstitial *interstitialAd;
@property (nonatomic, assign) BOOL showAdWhenLoaded;
@property (nonatomic, strong) UserDefaultsSettings *settings;
@end

@implementation InterstitialViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.settings = [UserDefaultsSettings sharedSettings];
    
    // Create a vertical stack for buttons
    UIStackView *buttonStack = [[UIStackView alloc] init];
    buttonStack.axis = UILayoutConstraintAxisVertical;
    buttonStack.spacing = 16;
    buttonStack.alignment = UIStackViewAlignmentCenter;
    buttonStack.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:buttonStack];
    
    // Load Interstitial button
    UIButton *loadButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [loadButton setTitle:@"Load Interstitial" forState:UIControlStateNormal];
    [loadButton addTarget:self action:@selector(loadInterstitialAd) forControlEvents:UIControlEventTouchUpInside];
    loadButton.backgroundColor = [UIColor systemGreenColor];
    [loadButton setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    loadButton.titleLabel.font = [UIFont boldSystemFontOfSize:16];
    loadButton.layer.cornerRadius = 8;
    loadButton.translatesAutoresizingMaskIntoConstraints = NO;
    [buttonStack addArrangedSubview:loadButton];
    
    // Show Interstitial button
    UIButton *showButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [showButton setTitle:@"Show Interstitial" forState:UIControlStateNormal];
    [showButton addTarget:self action:@selector(showInterstitialAd) forControlEvents:UIControlEventTouchUpInside];
    showButton.backgroundColor = [UIColor systemBlueColor];
    [showButton setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    showButton.titleLabel.font = [UIFont boldSystemFontOfSize:16];
    showButton.layer.cornerRadius = 8;
    showButton.translatesAutoresizingMaskIntoConstraints = NO;
    [buttonStack addArrangedSubview:showButton];
    
    // Button constraints
    [NSLayoutConstraint activateConstraints:@[
        [buttonStack.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor],
        [buttonStack.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor constant:100],
        [loadButton.widthAnchor constraintEqualToConstant:200],
        [loadButton.heightAnchor constraintEqualToConstant:44],
        [showButton.widthAnchor constraintEqualToConstant:200],
        [showButton.heightAnchor constraintEqualToConstant:44]
    ]];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    // No auto-loading - user must press Load Interstitial button
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    // Only reset if we're actually leaving the tab — not when a fullscreen ad
    // is presented over us (which also triggers viewWillDisappear).
    if (!self.interstitialAd || self.isLoading) {
        [self resetAdState];
    }
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (NSString *)adUnitId {
    return [[CLXDemoConfigManager sharedManager] currentConfig].interstitialAdUnitId;
}

- (void)loadInterstitialAd {
    
    if (self.isLoading) {
        [self showAlertWithTitle:@"Info" message:@"Interstitial is already loading."];
        return;
    }
    
    if (self.interstitialAd) {
        [self showAlertWithTitle:@"Info" message:@"Interstitial already loaded. Use Show Interstitial to display it."];
        return;
    }
    
    [self loadInterstitial];
}

- (void)loadInterstitial {

    if (self.isLoading || self.interstitialAd) {
        return;
    }

    self.isLoading = YES;
    [self updateStatusUIWithState:AdStateLoading];

    // Get ad unit ID from config, with settings override if provided
    NSString *adUnitId = [self adUnitId];
    if (_settings.interstitialAdUnitId.length > 0) {
        adUnitId = _settings.interstitialAdUnitId;
    }
    
    self.interstitialAd = [[CloudXCore shared] createInterstitialWithAdUnitId:adUnitId];
    self.interstitialAd.delegate = self;
    self.interstitialAd.revenueDelegate = self;
    
    if (self.interstitialAd) {
        [self.interstitialAd load];
    } else {
        self.isLoading = NO;
        [self updateStatusUIWithState:AdStateNoAd];
        [self showAlertWithTitle:@"Error" message:@"Failed to create interstitial."];
    }
}


- (void)showInterstitialAd {
    
    if (!self.interstitialAd) {
        [self showAlertWithTitle:@"Error" message:@"No interstitial loaded. Please load an interstitial first."];
        return;
    }
    
    if (self.isLoading) {
        [self showAlertWithTitle:@"Info" message:@"Interstitial is still loading. Please wait."];
        return;
    }
    
    if (self.interstitialAd.isReady) {
        [self.interstitialAd showFromViewController:self
                                          placement:@"demo_interstitial"
                                         customData:@"level:5,coins:100"];
    } else {
        [self showAlertWithTitle:@"Error" message:@"Interstitial is not ready. Please try loading again."];
    }
}

- (void)resetAdState {
    self.interstitialAd = nil;
    self.isLoading = NO;
    self.showAdWhenLoaded = NO;
    [self updateStatusUIWithState:AdStateNoAd];
}

#pragma mark - CLXInterstitialDelegate

- (void)didLoadAd:(CLXAd *)ad {
    [[DemoAppLogger sharedInstance] logAdEvent:@"✅ Interstitial didLoadAd" ad:ad];
    self.isLoading = NO;
    [self updateStatusUIWithState:AdStateReady];
}

- (void)didFailToLoadAd:(NSString *)adUnitId error:(CLXError *)error {
    [[DemoAppLogger sharedInstance] logMessage:[NSString stringWithFormat:@"❌ Interstitial failed to load (%@) - Error: %@", adUnitId, error ? error.localizedDescription : @"Unknown error"]];
    self.isLoading = NO;
    [self updateStatusUIWithState:AdStateNoAd];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        NSString *errorMessage = error ? [error detailedDemoDescription] : @"Unknown error occurred";
        [self showAlertWithTitle:@"Interstitial Ad Load Failed" message:errorMessage];
        self.interstitialAd = nil;
    });
}

- (void)didDisplayAd:(CLXAd *)ad {
    [[DemoAppLogger sharedInstance] logAdEvent:@"👀 Interstitial didDisplayAd" ad:ad];
}

- (void)didFailToDisplayAd:(CLXAd *)ad error:(CLXError *)error {
    [[DemoAppLogger sharedInstance] logAdEvent:@"❌ Interstitial didFailToDisplayAd" ad:ad];
    [self updateStatusUIWithState:AdStateNoAd];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        self.interstitialAd = nil;
        NSString *errorMessage = error ? [error detailedDemoDescription] : @"Unknown error occurred";
        [self showAlertWithTitle:@"Interstitial Ad Show Failed" message:errorMessage];
    });
}

- (void)didHideAd:(CLXAd *)ad {
    [[DemoAppLogger sharedInstance] logAdEvent:@"🔚 Interstitial didHideAd" ad:ad];
    
    self.showAdWhenLoaded = NO;
    self.interstitialAd = nil;
    
    // Don't auto-load - user must press Load Interstitial button
    [self updateStatusUIWithState:AdStateNoAd];
}

- (void)didClickAd:(CLXAd *)ad {
    [[DemoAppLogger sharedInstance] logAdEvent:@"👆 Interstitial didClickAd" ad:ad];
}

- (void)didPayRevenueForAd:(CLXAd *)ad {
    [[DemoAppLogger sharedInstance] logAdEvent:@"💰 Interstitial didPayRevenue" ad:ad];
}

@end 

File: demo-app-swift/CloudXSwiftRemotePods/Views/Ad/InterstitialViewController.swift
import UIKit
import CloudXCore

class InterstitialViewController: BaseAdViewController {
    private var interstitialAd: CLXInterstitial?
    private var showAdWhenLoaded = false
    private let settings = UserDefaultsSettings.shared
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setupUI()
        setupNotifications()
        updateStatusUI(state: AdState.noAd)
    }
    
    private func setupUI() {
        // Create a vertical stack for buttons
        let buttonStack = UIStackView()
        buttonStack.axis = .vertical
        buttonStack.spacing = 16
        buttonStack.alignment = .center
        buttonStack.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(buttonStack)
        
        // Load Interstitial button
        let loadButton = UIButton(type: .system)
        loadButton.setTitle("Load Interstitial", for: .normal)
        loadButton.addTarget(self, action: #selector(loadInterstitialAd), for: .touchUpInside)
        loadButton.backgroundColor = .systemGreen
        loadButton.setTitleColor(.white, for: .normal)
        loadButton.titleLabel?.font = .boldSystemFont(ofSize: 16)
        loadButton.layer.cornerRadius = 8
        loadButton.translatesAutoresizingMaskIntoConstraints = false
        buttonStack.addArrangedSubview(loadButton)
        
        // Show Interstitial button
        let showButton = UIButton(type: .system)
        showButton.setTitle("Show Interstitial", for: .normal)
        showButton.addTarget(self, action: #selector(showInterstitialAd), for: .touchUpInside)
        showButton.backgroundColor = .systemBlue
        showButton.setTitleColor(.white, for: .normal)
        showButton.titleLabel?.font = .boldSystemFont(ofSize: 16)
        showButton.layer.cornerRadius = 8
        showButton.translatesAutoresizingMaskIntoConstraints = false
        buttonStack.addArrangedSubview(showButton)
        
        // Button constraints
        NSLayoutConstraint.activate([
            buttonStack.centerXAnchor.constraint(equalTo: view.centerXAnchor),
            buttonStack.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 100),
            loadButton.widthAnchor.constraint(equalToConstant: 200),
            loadButton.heightAnchor.constraint(equalToConstant: 44),
            showButton.widthAnchor.constraint(equalToConstant: 200),
            showButton.heightAnchor.constraint(equalToConstant: 44)
        ])
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        // No auto-loading - user must press Load Interstitial button
    }
    
    @objc private func loadInterstitialAd() {
        receivedCallbacks = []
        if isLoading {
            showAlert(title: "Info", message: "Interstitial is already loading.")
            return
        }
        
        if interstitialAd != nil {
            showAlert(title: "Info", message: "Interstitial already loaded. Use Show Interstitial to display it.")
            return
        }
        
        loadInterstitial()
    }
    
    private func loadInterstitial() {

        if isLoading || interstitialAd != nil {
            return
        }

        isLoading = true
        updateStatusUI(state: AdState.loading)

        // Get adUnitId from config manager (with settings override if provided)
        var adUnitId = CLXDemoConfigManager.sharedManager.currentConfig.interstitialAdUnitId
        if !settings.interstitialAdUnitId.isEmpty {
            adUnitId = settings.interstitialAdUnitId
        }
        
        interstitialAd = cloudX.createInterstitial(adUnitId: adUnitId)
        interstitialAd?.delegate = self
        interstitialAd?.revenueDelegate = self
        
        if let interstitialAd = interstitialAd {
            interstitialAd.load()
        } else {
            isLoading = false
            updateStatusUI(state: AdState.noAd)
            showAlert(title: "Error", message: "Failed to create interstitial.")
        }
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        // Only reset if we're actually leaving the tab — not when a fullscreen ad
        // is presented over us (which also triggers viewWillDisappear).
        if interstitialAd == nil || isLoading {
            resetAdState()
        }
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
    
    private enum AdState {
        case noAd
        case loading
        case ready
        
        var text: String {
            switch self {
            case .noAd: return "No Ad Loaded"
            case .loading: return "Loading Ad..."
            case .ready: return "Ad Ready"
            }
        }
        
        var color: UIColor {
            switch self {
            case .noAd: return .systemRed
            case .loading: return .systemYellow
            case .ready: return .systemGreen
            }
        }
    }
    
    private func updateStatusUI(state: AdState) {
        DispatchQueue.main.async { [weak self] in
            self?.statusLabel.text = state.text
            self?.statusLabel.textColor = state.color
            self?.statusIndicator.backgroundColor = state.color
        }
    }
    
    private func setupNotifications() {
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleSDKInitialized),
            name: .sdkInitialized,
            object: nil
        )
    }
    
    @objc private func handleSDKInitialized() {
        // Don't auto-create - wait for user to press Load Interstitial button
    }
    
    // Remove the old createInterstitialAd and polling methods - they're replaced by loadInterstitial
    
    @objc private func showInterstitialAd() {
        
        guard let interstitialAd = interstitialAd else {
            showAlert(title: "Error", message: "No interstitial loaded. Please load an interstitial first.")
            return
        }
        
        if isLoading {
            showAlert(title: "Info", message: "Interstitial is still loading. Please wait.")
            return
        }
        
        if interstitialAd.isReady {
            interstitialAd.show(from: self, placement: "demo_interstitial", customData: "level:5,coins:100")
        } else {
            showAlert(title: "Error", message: "Interstitial is not ready. Please try loading again.")
        }
    }
    
    private func resetAdState() {
        interstitialAd = nil
        isLoading = false
        showAdWhenLoaded = false
        receivedCallbacks = []
        updateStatusUI(state: AdState.noAd)
    }
}

extension InterstitialViewController: CLXInterstitialDelegate, CLXAdRevenueDelegate {
    func didLoad(_ ad: CLXAd) {
        DemoAppLogger.sharedInstance.logAdEvent("✅ Interstitial didLoadAd", ad: ad)
        isLoading = false
        updateStatusUI(state: AdState.ready)
        // Don't auto-show - wait for user to press Show Interstitial button
    }
    
    func didFailToLoadAd(_ adUnitId: String, error: CLXError) {
        DemoAppLogger.sharedInstance.logMessage("❌ Interstitial failed to load (\(adUnitId)) - Error: \(error.localizedDescription)")
        isLoading = false
        updateStatusUI(state: AdState.noAd)
        
        DispatchQueue.main.async { [weak self] in
            let errorMessage = (error as NSError).detailedDemoDescription
            self?.showAlert(title: "Interstitial Ad Load Failed", message: errorMessage)
            self?.interstitialAd = nil
        }
    }
    
    func didDisplay(_ ad: CLXAd) {
        DemoAppLogger.sharedInstance.logAdEvent("👀 Interstitial didDisplayAd", ad: ad)
    }
    
    func didFailToDisplay(_ ad: CLXAd, error: CLXError) {
        DemoAppLogger.sharedInstance.logAdEvent("❌ Interstitial didFailToDisplayAd", ad: ad)
        updateStatusUI(state: AdState.noAd)
        
        DispatchQueue.main.async { [weak self] in
            self?.interstitialAd = nil
            let errorMessage = (error as NSError).detailedDemoDescription
            self?.showAlert(title: "Interstitial Ad Show Failed", message: errorMessage)
        }
    }
    
    func didHide(_ ad: CLXAd) {
        DemoAppLogger.sharedInstance.logAdEvent("🔚 Interstitial didHideAd", ad: ad)
        
        showAdWhenLoaded = false
        interstitialAd = nil
        
        // Don't auto-load - user must press Load Interstitial button
        updateStatusUI(state: AdState.noAd)
    }
    
    func didClick(_ ad: CLXAd) {
        receivedCallbacks.insert(.clicked)
        DemoAppLogger.sharedInstance.logAdEvent("👆 Interstitial didClickAd", ad: ad)
    }
    
    func didPayRevenue(for ad: CLXAd) {
        receivedCallbacks.insert(.revenueReceived)
        DemoAppLogger.sharedInstance.logAdEvent("💰 Interstitial didPayRevenue", ad: ad)
    }
} 
